pkg/template/resource.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2019 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package template
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	pipelinev1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+)
+
+// AddDefaultParamsFromSpec returns the params with the addition of all
+// paramSpecs that have default values and are already in the params list
+func AddDefaultParamsFromSpecs(params []pipelinev1.Param, paramSpecs []pipelinev1.ParamSpec) []pipelinev1.Param {
+	allParamsMap := map[string]pipelinev1.ArrayOrString{}
+	for _, paramSpec := range paramSpecs {
+		if paramSpec.Default != nil {
+			allParamsMap[paramSpec.Name] = *paramSpec.Default
+		}
+	}
+	for _, param := range params {
+		allParamsMap[param.Name] = param.Value
+	}
+	allParams := make([]pipelinev1.Param, len(allParamsMap))
+	i := 0
+	for name, value := range allParamsMap {
+		allParams[i] = pipelinev1.Param{Name: name, Value: value}
+		i++
+	}
+	return allParams
+}
```

Comment:
often it turns out to be better in the longrun to separate mutating an object from any other logic, so something like:

- Get all the params with their provided values
- Get all the params with default values
- Merge the two

Or maybe that's pretty much what's happening and a different name might make that clear, e.g. `MergeInDefaultParams`

Reply:
Yes, I think `MergeInDefaultParams` would be a better name.

Are you also suggesting that I split this function into two functions? One that gets all the params with default values (turns `[]pipelinev1.ParamSpec` into `[]pipelinev1.Param`). Then the second which merges the `params []pipelinev1.Param` and `defaultParams []pipelinev1.Param`?